src/Transition.js

```diff
@@ -496,31 +512,45 @@ Transition.propTypes = {
   /**
    * Callback fired after the "entered" status is applied. An extra parameter
    * `isAppearing` is supplied to indicate if the enter stage is occurring on the initial mount
+   * Note: when `nodeRef` prop is passed, `node` is not passed
    *
    * @type Function(node: HtmlElement, isAppearing: bool) -> void
    */
   onEntered: PropTypes.func,
 
   /**
    * Callback fired before the "exiting" status is applied.
+   * Note: when `nodeRef` prop is passed, `node` is not passed
    *
    * @type Function(node: HtmlElement) -> void
    */
   onExit: PropTypes.func,
 
   /**
    * Callback fired after the "exiting" status is applied.
+   * Note: when `nodeRef` prop is passed, `node` is not passed
    *
    * @type Function(node: HtmlElement) -> void
    */
   onExiting: PropTypes.func,
 
   /**
    * Callback fired after the "exited" status is applied.
+   * Note: when `nodeRef` prop is passed, `node` is not passed
    *
    * @type Function(node: HtmlElement) -> void
    */
   onExited: PropTypes.func,
+
+  /**
+   * A react reference to DOM element that need to transition
+   * https://stackoverflow.com/a/51127130/4671932
+   * Be aware when changing `key` prop of `Transition` in a `TransitionGroup`
+   * a new `nodeRef` need to be provided to `Transition` with changed `key` prop
+   * e.g. test/CSSTransition-test.js
+   * CSSTransition > reentering > should remove dynamically applied classes
+   */
+  nodeRef: PropTypes.shape({ current: PropTypes.instanceOf(Element) })
```

Comment:
I suggest adding this as the first prop, that way we'll encourage people to use it and we'll immediately let people know that callbacks will receive the `node` argument only if `nodeRef` isn't defined. There will still be risk of people expecting the `node` argument, but it will be lower.